Simplifying DBOS TypeScript Syntax

chuck-dbos opened the pull request:
**Main Change - New API**
In addition to supporting all prior API constructs, DBOS Transact Typescript has a new API that does not require passing contexts around between functions.  (This was inspired by the relative simplicity of the Python API.)  Instead, functionality is accessed by `DBOS.<thing>`, such as `@DBOS.workflow` or `DBOS.knexClient` or `DBOS.authenticatedUser`.  You can mix old and new code.

**Second Big Change - Library Usage**
DBOS can now be used as a library, or within Express.js or other frameworks, rather than just as serverless Koa-based entrypoints.  Examples and guidance are evolving.

**Other Items**
Recovery has been moved to after init.

maxdml commented on src/dbos-executor.ts:
```diff
@@ -691,7 +711,17 @@ export class DBOSExecutor implements DBOSExecutorContext {
-        result = await wf.call(params.configuredInstance, wCtxt, ...args);
+        let cresult: R | undefined;
+        await runWithWorkflowContext(wCtxt, async ()=> {
+          if (params.usesContext ?? true) {
```
why do we need to use `??` here? `params.usesContext` is a boolean

maxdml commented on src/context.ts:
```diff
@@ -3,9 +3,130 @@ import { GlobalLogger as Logger, Logger as DBOSLogger } from "./telemetry/logs";
+
+export async function runWithTransactionContext<Client extends UserDatabaseClient, R>(ctx: TransactionContextImpl<Client>, callback: ()=>Promise<R>) {
+  // Check we are in a workflow context and not in a step / transaction already
+  const pctx = getCurrentContextStore();
```
what does `pctx` stands for?

chuck-dbos replied:
parent / present

maxdml commented on src/context.ts:
```diff
@@ -3,9 +3,130 @@ import { GlobalLogger as Logger, Logger as DBOSLogger } from "./telemetry/logs";
+  callback);
+}
+
+export async function runWithStepContext<R>(ctx: StepContextImpl, callback: ()=>Promise<R>) {
+  // Check we are in a workflow context and not in a step / transaction already
+  const pctx = getCurrentContextStore();
+  if (!pctx) throw new DBOSInvalidWorkflowTransitionError();
+  if (!isInWorkflowCtx(pctx)) throw new DBOSInvalidWorkflowTransitionError();
+
+  return await asyncLocalCtx.run({
+    ctx,
+    workflowId: ctx.workflowUUID,
+    curStepFunctionId: ctx.functionID,
+    parentCtx: pctx,
+  },
+  callback);
+}
```
Who are these "run with X context" functions intended for?

chuck-dbos replied:
Intended for?  Or used for?  That one is used to run the callback function within availability of the StepContext.  Since we know it is a step context (and we're invoking a step), we can pick things out of it that may not have been possible w/ the base interface.

maxdml commented on src/data_validation.ts:
```diff
@@ -1,20 +1,23 @@
     const validationError = (msg:string) => {
         const err = new DBOSDataValidationError(msg);
-        opCtx?.span.addEvent("DataValidationError", { message: err.message });
+        opCtx?.span?.addEvent("DataValidationError", { message: err.message });
```
When can this span be missing?

maxdml replied:
I see. But in practice, given how we use `validationError`, this should never happen (we always use `validationError` for arguments number 2 and more, and `opCtx` is set either when entering `validateMethodArgs`, either when checking the first argument.)

maxdml commented on src/data_validation.ts:
```diff
@@ -1,20 +1,23 @@
     // Input validation
     methReg.args.forEach((argDescriptor, idx) => {
-        if (idx === 0)
+        if (idx === 0 && methReg.passContext)
```
Just to understand: what this is doing is gating method arguments validation to operations registered with a context. And this is because we don't have --- and don't see huge benefit in having -- a context free version of `getOrCreateMethodArgsRegistration` ?

chuck-dbos replied:
We're sharing the wrapper.  It needs to know if the context is there or not.  If we did split it we'd be careful not to do too much cut-n-paste?

maxdml commented on src/debugger/debug_workflow.ts:
```diff
@@ -173,7 +173,12 @@ export class WorkflowContextDebug extends DBOSContextImpl implements WorkflowCon
       }
       // If we have a proxy, then execute the user's transaction.
-      const result = await txn.call(clsinst, tCtxt, ...args);
+      let cresult: R | undefined;
+      await runWithTransactionContext(tCtxt, async ()=> {
+        cresult = await txn.call(clsinst, tCtxt, ...args);
+      });
+      const result = cresult!;
```
So what this does is enforcing the different asserts in `runWithTransactionContext`, right? Because otherwise, we're already calling the `txn` function with the "current" context (`tCtxt`)

chuck-dbos replied:
I may not understand the question.  I want to ensure that the same ctx is available that is passed in as the argument, that's the invariant. tCtxt was made up just there, so it ought to be in place...

maxdml commented on src/dbos-executor.ts:
```diff
@@ -647,6 +667,30 @@ export class DBOSExecutor implements DBOSExecutorContext {
+      params.parentCtx.assumedRole = pctx.assumedRole || "";
+      params.parentCtx.authenticatedRoles = pctx.authenticatedRoles || [];
+      params.parentCtx.workflowUUID = workflowUUID;
+    }
```
@chuck-dbos If the workflow is a child workflow, I think we'll not need to do this custom extraction. This is only useful when the workflow is a "root" workflow, `params.parentCtx` is not set, and we have to fill it manually to have a full `WorkflowContextImpl`. Can I just add a definition check on `params.parentCtx` in the `if` condition gating this logic?

chuck-dbos replied:
Didn't you (or copilot) write that?  I thought you knew ;-)

maxdml commented on src/dbos.ts:
```diff
@@ -0,0 +1,781 @@
+          return await cwfh.getResult();
+        }
+
+        const wfParams: InternalWorkflowParams = {
+          workflowUUID: wfId,
+          queueName: pctx?.queueAssignedForWorkflows,
+          configuredInstance : inst
```
We could decide to do all the `pctx` introspection here and create a new `WorkflowContextImpl` instance, configured with all we want, and set it as parentCtx. I think this would be cleaner than the current logic in dbos-executor.

maxdml commented on src/dbos-executor.ts:
```diff
@@ -647,6 +667,29 @@ export class DBOSExecutor implements DBOSExecutorContext {
+        wf.name,
+        {
+          operationUUID: workflowUUID,
+          operationType: OperationType.WORKFLOW,
```
This operationType, should be defined by the workflow "parent". So for instance, HTTP handler.

maxdml commented on src/dbos-executor.ts:
```diff
@@ -831,6 +886,135 @@ export class DBOSExecutor implements DBOSExecutorContext {
+        }
+        else {
+          await runWithTransactionContext(tCtxt, async ()=> {
+            const tf = txn as unknown as (...args: T)=>Promise<R>;
```
Should we define a `ContextFreeTransaction` type? :) Or overkill?

chuck-dbos replied:
I do not know why you would do that

maxdml commented on src/dbos-executor.ts:
```diff
@@ -973,6 +1294,7 @@ export class DBOSExecutor implements DBOSExecutorContext {
     }
     await this.scheduler?.destroyScheduler();
     wfQueueRunner.stop();
+    await this.wfqEnded;
```
That was a pre-existing bug right?

chuck-dbos replied:
I think the accounting just moved; had been in the runtime.

kraftp reviewed (commented):
Overall, this new syntax looks good! I think we can merge this reasonably soon as a first step as it's non-breaking then look at templates/docs.

A few questions:

- How does the server and HTTP decorators fit into this new system?  Is a server automatically launched by DBOS, or only if the HTTP decorators are present?  Can you completely ignore the existence of the HTTP component when building with Express/Nest/Next?
- How do entrypoints play into this?  What does DBOS do on launch now?

maxdml commented:
> How does the server and HTTP decorators fit into this new system? Is a server automatically launched by DBOS, or only if the HTTP decorators are present? Can you completely ignore the existence of the HTTP component when building with Express/Nest/Next?

DBOS.launch() will start an HTTP admin server. For the application HTTP component, you can ignore it completely. I don't think this change with what @chuck-dbos mentioned about HTTP decorators today.

> How do entrypoints play into this?

It depends on the application. If you use a TS transpiler that places the DBOS code in an "easy place" (i.e., not bundled with other chunks of code "somewhere", like next would do, you can still use the entrypoints. You can also simply import the DBOS classes before `DBOS.launch()` is called.

> What does DBOS do on launch now?

On the transact side, this depends on your app. DBOS.launch() must be in a piece of code that's executed when your program starts (e.g., `node dist/main.js`). On the cloud side, we'll need to add support for a start command in `dbos-config.yaml` (Otherwise we have to resort to hacks like `cp something node_modules/.bin/dbos-sdk`)

chuck-dbos commented:
> * How does the server and HTTP decorators fit into this new system?
You can call new functions from old, so you could use the old decorators.  I also added new ones like DBOS.getApi, and you can get the Koa context from DBOS.koaContext.

>  Is a server automatically launched by DBOS, or only if the HTTP decorators are present

I changed this today, the koa server is not supposed to be launched by DBOS if you don't use the decorators.

> Can you completely ignore the existence of the HTTP component when building with Express/Nest/Next?
It's supposed to (as of today).  However there's still the admin server.  Which I suppose you can ignore?  With more experience, I think we'll learn more about this.

> * How do entrypoints play into this?  What does DBOS do on launch now?

That's potentially an issue, I think Max is looking at it.  If you use the HTTP decorators, no issue with entrypoints.  It would also load the entrypoint file and run any code that's in there.  But if you opened a port, it wouldn't get passed through from load balancer.

maxdml commented on src/context.ts:
```diff
@@ -3,9 +3,130 @@ import { GlobalLogger as Logger, Logger as DBOSLogger } from "./telemetry/logs";
+
+export interface DBOSLocalCtx {
+  ctx?: DBOSContext;
+  parentCtx?: DBOSLocalCtx;
```
Are we ever using this `parentCtx` from other `DBOSLocalCtx` instances? I see how it is read from, e.g., `runWithStepContext`, but I don't see how it would be use with the ALS runner 🤔

qianl15 replied:
I think it's used in OTel tracing for tracking the parent spans.

kraftp reviewed (commented):
Like we discussed, my biggest remaining concern is startup commands when not using the DBOS runtime, especially in DBOS Cloud. Do we still need entrypoints? Or do we need a startup command like in Python? Or both?

maxdml commented:
> Do we still need entrypoints? Or do we need a startup command like in Python? Or both?

Both are useful

devhawk commented on packages/create/templates/hello-express/src/main.ts:
```diff
@@ -0,0 +1,16 @@
+  });
+}
+
+main().then(()=>{}).catch((e) => console.log(e));
```
```suggestion
main().catch(console.log);
```

kraftp commented on dbos-config.schema.json:
```diff
@@ -145,6 +145,13 @@
         "admin_port": {
             "type": "number",
             "description": "The port number of the admin server (Default: 3001 or the app server port + 1)"
+        },
```
Just to confirm, this is exactly the same as in Python?

maxdml replied:
The start command? Yes, I checked the Python implementation for reference. It
- Takes a list of commands
- Run each of them (blocking)

kraftp commented on packages/create/templates/hello-express/dbos-config.yaml:
```diff
@@ -0,0 +1,20 @@
+    - npx knex migrate:rollback
+runtimeConfig:
+  entrypoints:
+    - dist/main.js
```
This is the part I don't understand. Since the app is built with Express and doesn't use our HTTP decorators, shouldn't it use the new start command instead of an entrypoint?

chuck-dbos replied:
Start command did not exist when that line was written.

maxdml replied:
The entrypoints are used to inform Transact where to load classes for registrations

chuck-dbos replied:
You may have noticed that I cheated and took advantage of the fact that the entrypoint file is essentially "run".  This can be improved with the new facilities.

qianl15 replied:
Now with the start command, should we change this?

kraftp commented on packages/create/templates/hello-express/src/main.ts:
```diff
@@ -0,0 +1,16 @@
+  const ENV = process.env.NODE_ENV || 'development';
+
+  app.listen(PORT, () => {
+    console.log(`🚀 Server is running on http://localhost:${PORT}`);
```
Why do we need a separate `main.ts` and `operations.ts`?  If we use the new start command, can we combine these files?

chuck-dbos replied:
I left it as 2 partly so I could unit test.  There may be a different way of doing that... but keeping unit tests seemed like a good idea to me.

kraftp commented on packages/create/templates/hello-v2/src/operations.ts:
```diff
@@ -0,0 +1,64 @@
+// First, let's import DBOS
+import { DBOS } from "@dbos-inc/dbos-sdk";
+
+// Then, let's declare a type representing the "dbos_hello" database table
```
Can this just replace hello completely? We should either remove or upgrade the old templates...

chuck-dbos replied:
We were thinking at some point in time, but that we're not really ready to handle the external name use yet.  Once there is doc, etc., then getting rid of old hello makes sense.
This could be a week or two, but the PR is already pretty big and probably shouldn't hang out that long.

kraftp reviewed (approved):
I think this is ready to merge--the core functionality is fully working. @qianl15 should also review. Next step is templates + docs, then we can officially release it.

qianl15 commented on packages/create/templates/hello-express/src/main.ts:
```diff
@@ -0,0 +1,16 @@
+async function main() {
+  await DBOS.launch({expressApp: app});
+
+  const PORT = DBOS.runtimeConfig?.port ?? 9000;
```
Nit: The default port for Express.js is 3000. Why do we use 9000 here?

chuck-dbos replied:
What port does the DBOS Cloud forward?

qianl15 replied:
DBOS Cloud forwards to 3000 for node.js apps and 8000 for Python apps.

qianl15 commented on packages/create/templates/hello-express/src/operations.ts:
```diff
@@ -0,0 +1,81 @@
+// DBOS uses TypeScript decorators to automatically make your functions reliable, so they need to be static.
+export class Hello {
+  // This function greets a user and increments the greet count in the database.
+  // The @DBOS.transaction() decorator ensures that this function runs as a transaction, only once.
```
```suggestion
  // The @DBOS.transaction() decorator ensures that this function runs as a database transaction.
```
Removed "only once" because it sounds confusing.

chuck-dbos replied:
That was a copy+paste, will change both.

qianl15 commented on packages/create/templates/hello-v2/dbos-config.yaml:
```diff
@@ -0,0 +1,20 @@
+  rollback:
+    - npx knex migrate:rollback
+runtimeConfig:
+  entrypoints:
```
Same as above, should we change it to use the new start command?

chuck-dbos replied:
Not for this one, it still uses serverless entrypoints.
If you compare, the serverless version gives you a handful of features for free with less lines of code since you don't set up the server.

qianl15 commented on packages/create/templates/hello-v2/package.json:
```diff
@@ -0,0 +1,23 @@
+{
+  "name": "dbos-hello-koa",
```
Do we want to call this template `hello-koa`? We don't use the word `v2` anywhere else.

chuck-dbos replied:
It was supposed to be v2.  We do have a longer-term plan to make this into the main hello when it is ready.
I do not want to call it hello-koa.  If you want koa, use our nice serverless stuff that helps you get everything right.

qianl15 commented on packages/create/templates/hello-v2/src/operations.test.ts:
```diff
@@ -0,0 +1,38 @@
+   * Test the HTTP endpoint.
+   */
+  /*
+  test("test-endpoint", async () => {
```
Remove this or uncomment?

chuck-dbos replied:
It needs to be supported at some point.

qianl15 commented on packages/create/templates/hello-v2/src/operations.ts:
```diff
@@ -0,0 +1,64 @@
+// DBOS uses TypeScript decorators to automatically make your functions reliable, so they need to be static.
+export class Hello {
+  // Serve this function from HTTP GET requests at the /greeting endpoint with 'user' as a path parameter
+  @DBOS.getApi("/greeting/:user")
```
If we want to eventually get rid of `getApi` and other HTTP verbs, we should use Koa directly here.

chuck-dbos replied:
Disagree strongly.

qianl15 replied:
We can discuss this some other time -- maintaining a thin wrapper around Koa is pretty annoying. Remember the times we wanted to support CORS and preflight requests :/

chuck-dbos replied:
Agree with one part of the Koa point (features have to be added in the framework), strong disagree on the CORS part, baking that into your code is bad and should be managed by the deployment, which we can do better if we help you in a "serverless" way.

qianl15 commented on src/context.ts:
```diff
@@ -3,9 +3,132 @@ import { GlobalLogger as Logger, Logger as DBOSLogger } from "./telemetry/logs";
+  idAssignedForNextWorkflow?: string;
+  queueAssignedForWorkflows?: string;
+  parentWorkflowId?: string;
+  parentWorkflowFid?: number;
```
What is this? This field is not used anywhere.

maxdml commented on src/context.ts:
```diff
@@ -3,9 +3,132 @@ import { GlobalLogger as Logger, Logger as DBOSLogger } from "./telemetry/logs";
+  assumedRole?: string;
+  request?: HTTPRequest;
+  operationType?: string; // A custom helper for users to set a operation type of their choice. Intended for functions setting a pctx to run DBOS operations from.
+  operationCaller?: string; // This is made to pass through the operationName to DBOS contexts, and potentially the caller span name.
```
Can be set by the caller to whatever makes sense to them -- this is mostly to make traces easier to parse

qianl15 reviewed (commented):
A general question: does it still use Stage 2 decorators?

chuck-dbos commented:
> A general question: does it still use Stage 2 decorators?

Yeah, this is permits an incremental upgrade without having file-specific TS compiler settings.

Have they made a better decorator scheme for TS?  A year ago, the successor was neither better nor final.

qianl15 commented on src/dbos.ts:
```diff
@@ -0,0 +1,893 @@
+
+// Declare all the HTTP applications a user can pass to the DBOS object during launch()
+// This allows us to add a DBOS tracing middleware (extract W3C Trace context, set request ID, etc)
+export interface DBOSHttpApps {
```
Hmm, what if there are multiple apps? Do we allow multiple of those (e.g., both koaApp and expressApp)?

chuck-dbos replied:
@maxdml 
You probably should just do 1, or extract the routes and make a single handler tree, I don't think you'd have a lot of luck getting multiple ports through the LB right now.

qianl15 reviewed (approved):
Nice work on TS v2!
